### PR TITLE
Using the secondary render target to scale the entire map at once.

### PR DESCRIPTION
**Pull request by authenticate**
Hi,

I started to finish the implementation of the secondary render target map mode code.  Not everything is completely finished at the moment.  However, I wanted to push something to up to get some feedback and make sure everything was on the right track.

The initial images look pretty good.

I'll try and comment the code itself to highlight some of the important areas.

A few major things to note are:

The code to clamp the camera to boundaries of the map is still commented out.  When I get some more time, I need to revisit this and update how it scales.

Map mode should use a normal coordinate system similar to other modes in the game (Well, I think it does now.).  The concept of the "screen offset" where everything was a multiple of two off should no longer apply.  The multiple of two is now applied right before drawing.  So, the game logic code itself is no longer dependent on this offset.  I tried to remove some of the legacy code related to this.

Finally, I think the particle system definitions were written with the "screen offset" taken into account.  So, they appeared twice as big in map mode.  I updated the save point particle systems as a proof of concept to get them working with the new system.  If this is correct, the other map mode particle systems still need to updated.

Let me know what you think.

Thanks.

### CODE REVIEW

File: src/modes/map/map_mode.cpp [hunk 1]
```diff
@@ -391,6 +383,11 @@ void MapMode::Draw()
     VideoManager->PushState();
     VideoManager->SetStandardCoordSys();
     VideoManager->SetDrawFlags(VIDEO_BLEND, VIDEO_X_CENTER, VIDEO_Y_BOTTOM, 0);
+
+    VideoManager->EnableSecondaryRenderTarget();
```
Comment (authenticate) [inline src/modes/map/map_mode.cpp]: The compositing and scaling logic takes place in here.

File: src/modes/map/map_mode.cpp [hunk 2]
```diff
@@ -968,46 +978,50 @@ void MapMode::_UpdateMapFrame()
...
-    // Camera exceeds the bottom boundary of the map
-    else if(_map_frame.tile_y_start + TILES_ON_Y_AXIS >= _tile_supervisor->_num_tile_on_y_axis) {
-        _map_frame.tile_y_start = static_cast<int16_t>(_tile_supervisor->_num_tile_on_y_axis - TILES_ON_Y_AXIS);
-        _map_frame.tile_y_offset = vt_utils::FloorToFloatMultiple(2.0f, _pixel_length_y);
-        _map_frame.screen_edges.bottom = static_cast<float>(_object_supervisor->_num_grid_y_axis);
-        _map_frame.screen_edges.top = _map_frame.screen_edges.bottom - SCREEN_GRID_Y_LENGTH;
-        _map_frame.num_draw_y_axis = TILES_ON_Y_AXIS;
-        _camera_y_in_map_corner = true;
-    }
+    //
```
Comment (authenticate) [inline src/modes/map/map_mode.cpp]: This camera clamping code still needs updated.

File: src/modes/map/map_tiles.cpp [hunk 3]
```diff
@@ -327,32 +327,42 @@ void TileSupervisor::DrawLayers(const MapFrame *frame, const LAYER_TYPE &layer_t
...
+        //
+
         // We substract 0.5 horizontally and 1.0 vertically here
         // because the video engine will display the map tiles using their
         // top left coordinates to avoid a position computation flaw when specifying the tile
         // coordinates from the bottom center point, as the engine does for everything else.
         VideoManager->Move(GRID_LENGTH * (frame->tile_x_offset - 1.0f), GRID_LENGTH * (frame->tile_y_offset - 2.0f));
-        for(uint32_t y = static_cast<uint32_t>(frame->tile_y_start); y < y_end; ++y) {
-            for(uint32_t x = static_cast<uint32_t>(frame->tile_x_start); x < x_end; ++x) {
+        for(int32_t y = static_cast<int32_t>(frame->tile_y_start); y < y_end; ++y) {
```
Comment (authenticate) [inline src/modes/map/map_tiles.cpp]: These changes are a side effect of commenting out the camera clamping code.  They should be able to be reverted once the camera clamping code is fixed.

**Comment by authenticate** [comment]
The camera clamping code should be working again now.  It's getting closer.

File: src/modes/map/map_mode.cpp [hunk 4]
```diff
@@ -968,46 +978,53 @@ void MapMode::_UpdateMapFrame()
...
-    // Camera exceeds the right boundary of the map
-    else if(_map_frame.tile_x_start + TILES_ON_X_AXIS >= _tile_supervisor->_num_tile_on_x_axis) {
-        _map_frame.tile_x_start = static_cast<int16_t>(_tile_supervisor->_num_tile_on_x_axis - TILES_ON_X_AXIS);
-        _map_frame.tile_x_offset = vt_utils::FloorToFloatMultiple(1.0f, _pixel_length_x);
-        _map_frame.screen_edges.right = static_cast<float>(_object_supervisor->_num_grid_x_axis);
-        _map_frame.screen_edges.left = _map_frame.screen_edges.right - SCREEN_GRID_X_LENGTH;
+    // Camera exceeds the right boundary of the map.
+    else if (_map_frame.tile_x_start + TILES_ON_X_AXIS >= _tile_supervisor->_num_tile_on_x_axis + width_offset) {
+        _map_frame.tile_x_start = static_cast<int16_t>(_tile_supervisor->_num_tile_on_x_axis - TILES_ON_X_AXIS + width_offset);
+        _map_frame.tile_x_offset = vt_utils::CeilToFloatMultiple(1.0f, _pixel_length_x);
```
Comment (authenticate) [inline src/modes/map/map_mode.cpp]: I think this change is alright.  On the upper boundary cases, "ceil" correcting pushes the value towards the upper boundary.

**Comment by Bertram25** [comment]
Hi again @authenticate :)

I must say I wanted to use some free time to actually do that job since you kindly provided the api to make it happen much more easily but it seems my spare time is flying really fast those days.
So I apologize for not advancing much on my current Valyrian TODO list.

That said, I'll happily review your work asap. :)

**Comment by Bertram25** [comment]
Compile Warnings:
```bash
ValyriaTear/src/modes/map/map_tiles.cpp: In member function ‘void vt_map::private_map::TileSupervisor::DrawLayers(const vt_map::private_map::MapFrame*, const vt_map::private_map::LAYER_TYPE&)’:
ValyriaTear/src/modes/map/map_tiles.cpp:355:40: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
                     layer.tiles.size() > y &&
                                        ^
ValyriaTear/src/modes/map/map_tiles.cpp:356:48: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
                     layer.tiles.front().size() > x &&
                                                ^
```

**Comment by Bertram25** [comment]
One error spotted on the map_objects.cpp Light object creation code.

**Comment by Bertram25** [comment]
Ok, I tested it, nice work already! It's amazing how the characters blend better when the settings has no tiling glitches!

Here are the improvements needed I've seen so far:
- The parallax effect applied notably on cloud shadows in the village is broken. Try moving around and you'll notice it right away.
- Light object display is broken (due to error when loading.)
- The interaction icons, such as the bubble icons, the save points, ... and most importantly the characters are over-blurred. This is especially noticeable if you play at 800x600 compared to the master version.
What I "think" is happening that somehow the entire scene is downsized before being magnified again when being drawn, making that blur effect appear too much compared to before. (Note to self: What is the size of the secondary rendering target buffer already?)
I've not checked yet, but do try the game before and you'll notice it right away as well.
Here are screenshot taken at 800x600 as an evidence:
Before patch:
screenshot_9
After patch:
screenshot_10

Also, the village in-house ground has clipping problems when at  a res of <1024xH, maybe linked to this part, or not. To be checked later.

**Comment by authenticate** [comment]
Thanks for the updates.

I went through and fixed the small issues, the warnings and the missing light object.  When I get some more time, I will go through and fix the larger issues.

> (Note to self: What is the size of the secondary rendering target buffer already?)

If I remember correctly, it is equal to size of the swap chain.  So, if the game is 1024x800, the secondary render target is also 1024x800.  I made the assumption that we always want to zoom in or, said a different way, the map scale is always >= 1.  With this assumption, the secondary render target will always be large enough for the visible portion of the map.  If we ever wanted to zoom out, we would need a more elaborate mechanism to determine the size of the secondary render target.

Now, as the map tiles and other images are composited into the secondary render target, they are supposed to be drawn at the their native resolutions.  So, when they are scaled up, they should be no more or no less blurry than the original implementation.  However, there is definitely an error somewhere because this is not actually the case.

**Comment by Bertram25** [comment]
Thanks for the quick fixes.

> If we ever wanted to zoom out, we would need a more elaborate mechanism to determine the size of the secondary render target.

To be honest, I don't think we may ever want to zoom out as it doesn't bring anything in term of game experience IMHO, or I don't see interesting use of it. On the other hand, a more interesting feature could be to dynamically change the zoom (but it would still be from 2.0 to more), to put the accent on certain scene parts during dialogues. And even though, it has to be tried to see whether it actually gives something interesting. All that said, cleaning the code toward this is a good thing in any case, as it makes the code more agnostic and more scalable.

An another part, when dynamic zooms could be clearly interesting, is during battle skill use, where a zoom could happen on a character doing a special blow or something.
But again, all that is needs for the future and nothing important for now in term of player experience.

Cheers,

**Comment by Bertram25** [comment]
@authenticate I just realized:

> If I remember correctly, it is equal to size of the swap chain. So, if the game is 1024x800, the secondary render target is also 1024x800.

Then, if I'm correct, when playing at 800x600, you have a second rendering target using 800x600, which explains the over blur. With the current method, the one before the patch, the render target is always using a coord system of 1024x768 and probably down size it when rendering through the view point (to 800x600 in our case).

Could you ensure the second rendering buffer is at least 1024x768?

Cheers,

**Comment by authenticate** [comment]
> With the current method, the one before the patch, the render target is always using a coord system of 1024x768 and probably down size it when rendering through the view point (to 800x600 in our case).

You are correct.

I made some changes so now the render target is always 1024x768.  Once things are working and in a better state, we can change it to something more robust.

I have been testing exclusively at 1024x768.  This way everything is as simple as possible to fix the blur issue.  I have not had much luck with it though.  (As a side note, other resolutions will have different visual problems with this build.  However, those are minor and can get cleaned up after everything is no longer blurry.)

Using gDEBugger, I can confirm the tiles are composited into the render target correctly.  They do not appear blurry and are the correct size at this stage.  The blur seems to be introduced when the render target is drawn to the screen.  I have to think about it some more.  I am not sure conceptually why this process would yield different results then scaling up the tiles individually as they are drawn.  It is probably something simple that I am missing.

In addition, I forgot only map tiles are natively stored at 32x32 pixels.  Other map objects, trees for example, can have much larger native sizes.  So, in the final version, only the map tiles themselves should be composited and scaled up.  The downscaling required to composite larger images would cause a fidelity loss for most map objects.

**Comment by Bertram25** [comment]
>  Other map objects, trees for example, can have much larger native sizes.

I'm not sure to get it fully. The following image is stored at original size (not scaled up within the png): https://github.com/ValyriaTear/ValyriaTear/blob/master/data/entities/map/objects/tree_big1.png
Objects may indeed be scaled up when being loaded through image frame size parameters in the corresponding animation lua file. E.g: https://github.com/ValyriaTear/ValyriaTear/blob/master/data/entities/map/objects/clock1.lua (frame_width, frame_height).

Could this cause a problem? Otherwises, to me, every objects are rescaled for the map mode the same way. Could explain to some old, and naive code bear? ;)

**Comment by authenticate** [comment]
You are absolutely correct.  For some strange reason I thought something else was going on.  Sorry about the confusion.

**Comment by authenticate** [comment]
I am going to close this pull request.  I am stuck at the moment and need to revisit it.  I will leave the code on my fork for future reference.

**Comment by Bertram25** [comment]
> I am going to close this pull request. I am stuck at the moment and need to revisit it. I will leave the code on my fork for future reference.

Np at all! Sometimes giving such things time to think make them easier.

**Pull request closed without merging**